CONTRIBUTING.md

```diff
@@ -0,0 +1,513 @@
+
+Without being unnecessarily verbose, explain how the output is calculated
+from the inputs. Explain the side effects. Explain any restrictions on the inputs. Explain failure
+conditions, including when the function will panic, return an error
+or return an empty value.
+
+**Example:**
+
+```rust
+/// A contract that does not violate its read-only declarations is called
+/// *read-only correct*.
+impl<'a, 'b> ReadOnlyChecker<'a, 'b> {
+    /// Checks each top-level expression in `contract_analysis.expressions`
+    /// for read-only correctness.
+    ///
+    /// Returns successfully iff this function is read-only correct.
+    ///
+    /// # Errors
+    ///
+    /// - Returns CheckErrors::WriteAttemptedInReadOnly if there is a read-only
+    ///   violation, i.e. if some function marked read-only attempts to modify
+    ///   the chainstate.
+    pub fn run(&mut self, contract_analysis: &ContractAnalysis) -> CheckResult<()>
+```
+
+This comment is considered positive because it explains the contract of the function in pseudo-code. Someone who understands the constructs mentioned could, e.g., write a test for this method from this description.
+
+#### Comments on Implementations of Virtual Methods 
+
+Note that, if a function implements a virtual function on an interface, the comments should not
+repeat what was specified on the interface declaration. The comment should only add information specific to that implementation.
+
+#### Data Member Comments
+
+Each data member in a struct should have a comment describing what that member
+is, and what it is used for. Such comments are usually brief but should
+clear up any ambiguity that might result from having only the variable
+name and type.
+
+**Example:**
+
+```rust
+pub struct ReadOnlyChecker<'a, 'b> {
+    /// Mapping from function name to a boolean indicating whether
+    /// the function with that name is read-only.
+    /// This map contains all functions in the contract analyzed.
+    defined_functions: HashMap<ClarityName, bool>,
+```
+
+This comment is considered positive because it clarifies users might have about the content and role of this member. E.g., it explains that the `bool` indicates whether the function is *read-only*, whereas this cannot be gotten from the signature alone.
+
+#### Test Comments
+
+Each test should have enough comments to help an unfamiliar reader understand:
+
+1. what is conceptually being tested
+1. why a given answer is expected
+
+Sometimes this can be obvious without much comments, perhaps from the context,
+or because the test is very simple. Often though, comments are necessary.
+
+**Example:**
+
+```rust
+#[test]
+#[ignore]
+fn transaction_validation_integration_test() {
+    /// The purpose of this test is to check if the mempool admission checks
```

Comment:
Why are these comments inside the function instead of outside?
```rust
#[test]
 #[ignore]
/// The purpose of this test is...
fn transaction_validation_integration_test() {
```

Reply:
Fixed in 87fd39163